BUG: Fix Period and PeriodIndex support of combined offsets aliases

agraboso opened the pull request:
- [x] closes #13730
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Essentially, makes sure any `freq` string passed to a `Period` or `PeriodIndex` goes through [`_Period._maybe_convert_freq()`](https://github.com/pydata/pandas/blob/master/pandas/src/period.pyx#L682-L697), which calls [`to_offset()`](https://github.com/pydata/pandas/blob/master/pandas/tseries/frequencies.py#L389-L451), which is where the logic for combining aliases is.

All the examples in #13730 result in the correct output, and all existing tests pass. I have not written any new ones yet &mdash; I first wanted to get the opinion of the maintainers.

This PR builds on #13868 (without it, some existing tests fail).

agraboso committed:
BUG: Fix Period and PeriodIndex support of combined alias offsets

Closes GH13730.

jreback commented:
this seems to be on top of #13868 was that intentional? (usually we try to do independent PR's). if they are not its ok, but then should just put in one / indicate at the top that is the case.

@sinhrks

jreback commented:
can you run an asv on this, make sure looks good for perf.

jreback commented:
also would need to add tests from the issue. I dont' think we are actually testing this (and that's why it doesn't fail).

sinhrks commented on pandas/src/period.pyx:
```diff
@@ -1203,6 +1208,7 @@ class Period(_Period):
 
 def _ordinal_from_fields(year, month, quarter, day,
                          hour, minute, second, freq):
+    freq = Period._maybe_convert_freq(freq)
```
this doesn't needed, as ``_ordinal_from_field`` is only called from here (after 1130th line)?

- https://github.com/pydata/pandas/blob/5f524d61fd336b850d34f13d5ffb2b6136073f21/pandas/src/period.pyx#L1152

agraboso replied:
That's right. I was being overly cautious (maybe some other part of the code might need to call this in the future?) but I can remove it.

agraboso commented:
I ran the asv benchmarks. Here are the results:

```
    before     after       ratio
  [59f25575] [01c55da7]
+    3.73ms    16.80ms      4.51  reindex.frame_drop_dup_na_inplace.time_frame_drop_dup_na_inplace
+  898.72ms      3.12s      3.47  plotting.plot_andrews_curves.time_plot_andrews_curves
-   45.25ms    13.87ms      0.31  period.create_period_index_from_date_range.time_period_index
```
I have no experience whatsoever with these, so I don't know how to interpret the results above, but I find it perplexing that the two tests that run slower have absolutely nothing to do with time series...

sinhrks commented:
Thanks for the test. The slowness looks to be caused by regex search. Maybe we should use dict lookup like ``get_offset`` does (related to #4205, merging ``to_offset`` and ``get_offset`` to make fast/simple impl).

- https://github.com/pydata/pandas/blob/master/pandas/tseries/frequencies.py#L508

agraboso commented:
@sinhrks I'm confused. The asv test that has to do with periods is faster with this PR...

sinhrks commented:
Ah sorry, I looked 1st 2 (of course unrelated to the period). Pls ignore.

jreback commented:
pls rebase. Need tests to close #13730 ; and add to whatsnew.

agraboso commented:
Rebased. I have also added tests for `Period`, `Period.asfreq`, `PeriodIndex`, `PeriodIndex.asfreq`, `PeriodIndex.to_datetime` and `period_range` with combined aliases.

There is one standing issue: the function [`get_standard_freq()`](https://github.com/agraboso/pandas/blob/fix-13730/pandas/tseries/frequencies.py#L541-L552) does not accepted combined aliases as of now. Apart from tests, it is only called from three places in the whole codebase:

- Lines [period.py#L609](https://github.com/agraboso/pandas/blob/fix-13730/pandas/tseries/period.py#L609) and [period.pyx#L742](https://github.com/agraboso/pandas/blob/fix-13730/pandas/src/period.pyx#L742) both read

    ```
    freqstr = frequencies.get_standard_freq(other)
    ```

    They can &mdash; and should &mdash; be changed to

    ```
    freqstr = other.rule_code
    ```

- [period.py#L488-L489](https://github.com/agraboso/pandas/blob/fix-13730/pandas/tseries/period.py#L488-L489)

    ```
    freq = Period._maybe_convert_freq(freq)
    freq = frequencies.get_standard_freq(freq)
    ```

    can &mdash; and should &mdash; be changed to

    ```
    freq = Period._maybe_convert_freq(freq).freqstr
    ```

I have tried these changes in my local copy of the repo and all tests pass.

In view of this, I propose deprecating `get_standard_freq()` (as is already done with [`get_offset_name()`](https://github.com/agraboso/pandas/blob/fix-13730/pandas/tseries/frequencies.py#L527-L538)) &mdash; or removing it completely if possible.

agraboso commented:
I implemented the deprecation proposal in the last commit.

sinhrks commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -864,3 +865,4 @@ Bug Fixes
 - Bug in ``.to_excel()`` when DataFrame contains a MultiIndex which contains a label with a NaN value (:issue:`13511`)
 - Bug in ``pd.read_csv`` in Python 2.x with non-UTF8 encoded, multi-character separated data (:issue:`3404`)
 - Bug in ``Index`` raises ``KeyError`` displaying incorrect column when column is not in the df and columns contains duplicate values (:issue:`13822`)
+- Bug in ``Period`` and ``PeriodIndex`` creating wrong dates when frequency has multiple offsets (:issue:`13874`)
```
"combined offsets" or "combined aliases" should be better. 

- http://pandas.pydata.org/pandas-docs/stable/timeseries.html#combining-aliases

agraboso commented:
Is "combined offset aliases" okay?

sinhrks commented:
@agraboso Thx, it sounds ok.

sinhrks commented on pandas/tseries/period.py:
```diff
@@ -478,7 +485,7 @@ def asfreq(self, freq=None, how='E'):
         how = _validate_end_alias(how)
 
-        freq = frequencies.get_standard_freq(freq)
+        freq = Period._maybe_convert_freq(freq).freqstr
```
Looks better to pass freq as it is, because ``_gfc`` can accept offset?

- https://github.com/pydata/pandas/blob/1f55e919a3b85cd0891a7a7a1d6913ca06f6bd9e/pandas/tseries/frequencies.py#L259

agraboso replied:
Do you mean stripping the `.freqstr` at the end? Sure, that works (tests pass)

sinhrks replied:
yes, thx to confirm.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -684,6 +684,7 @@ Deprecations
 - ``pd.tseries.util.pivot_annual`` is deprecated. Use ``pivot_table`` as alternative, an example is :ref:`here <cookbook.pivot>` (:issue:`736`)
 - ``pd.tseries.util.isleapyear`` has been deprecated and will be removed in a subsequent release. Datetime-likes now have a ``.is_leap_year`` property. (:issue:`13727`)
 - ``Panel4D`` and ``PanelND`` constructors are deprecated and will be removed in a future version. The recommended way to represent these types of n-dimensional data are with the `xarray package <http://xarray.pydata.org/en/stable/>`__. Pandas provides a :meth:`~Panel4D.to_xarray` method to automate this conversion. (:issue:`13564`)
+- ``pd.tseries.frequencies.get_standard_freq`` is deprecated. Use `freq.rule_code` if `freq` is a `DateOffset` object, or `pd.tseries.frequencies.to_offset(freq).rule_code` if it is a string or tuple. (:issue:`13874`)
```
was this ever an advertised function? I think we can just remove it.

jreback replied:
change to match revised msg below

jreback replied:
write it out as: ``pandas.tseries.frequencies.get_standard_freq``

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -542,14 +542,16 @@ def get_standard_freq(freq):
-        return None
+
+    msg = ("get_standard_freq(freq) is deprecated. "
+           "Use freq.rule_code instead if freq is a DateOffset object, "
```
just recommend ``.to_offset``. (which accepts all of these). BTW, can we enhance the doc-string of ``.to_offset``? (add Parameters / more examples)

agraboso commented:
@jreback How does it look?

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,76 @@ def get_period_alias(offset_str):
+
+    Parameters
+    ----------
+    freq : str or tuple or datetime.timedelta or DateOffset or None
```
use commas rather than or

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,76 @@ def get_period_alias(offset_str):
+    Parameters
+    ----------
+    freq : str or tuple or datetime.timedelta or DateOffset or None
+
```
add a Raises section

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,76 @@ def get_period_alias(offset_str):
+    -------
+    delta : DateOffset
+        None if freq is None
+        ValueError if freq is an invalid frequency
```
e.g. this ValueError goes in Raises

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,76 @@ def get_period_alias(offset_str):
+
+    Returns
+    -------
+    delta : DateOffset
```
can also add a see also to DateOffset

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,76 @@ def get_period_alias(offset_str):
 
 
-def to_offset(freqstr):
+def to_offset(freq):
```
this is technically a change, e.g. someone passes by name.

so we can either change it back or deprecate it (deprecating it is fine). you can use the ``deprecate`` decorator

agraboso commented:
@jreback Done.

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -386,51 +386,85 @@ def get_period_alias(offset_str):
+
+    See Also
+    --------
+    DateOffset
```
make this ``pandas.DateOffset``

jreback commented:
you are going to have a flake8 minor issue (long lines). otherwise lgtm.

can you run an asv vs master (just ``-b period``) is fine, and show the results.

agraboso commented:
> you are going to have a flake8 minor issue (long lines). otherwise lgtm.

`flake8` does not complain. Which are the long lines you refer to?

> can you run an asv vs master (just -b period) is fine, and show the results.

```
    before     after       ratio
  [768bf495] [81262a57]
-   54.20ms    14.96ms      0.28  period.create_period_index_from_date_range.time_period_index
```

jreback commented:
are u just showing the single change in asv?
there are more period benchmarks

agraboso commented:
Here is the whole output:

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Installing into conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 24 total benchmarks (2 commits * 1 environments * 12 benchmarks)
[  0.00%] · For pandas commit hash 81262a57:
[  0.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....................................
[  0.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  4.17%] ··· Running gil.nogil_datetime_fields.time_datetime_to_period                                                                                                    18.29ms
[  8.33%] ··· Running gil.nogil_datetime_fields.time_period_to_datetime                                                                                                    55.16ms
[ 12.50%] ··· Running period.create_period_index_from_date_range.time_period_index                                                                                         14.96ms
[ 16.67%] ··· Running period.period_algorithm.time_period_index_drop_duplicates                                                                                            30.93μs
[ 20.83%] ··· Running period.period_algorithm.time_period_index_value_counts                                                                                                1.12ms
[ 25.00%] ··· Running period.period_algorithm.time_period_series_drop_duplicates                                                                                           16.60ms
[ 29.17%] ··· Running period.period_algorithm.time_period_series_value_counts                                                                                               6.94ms
[ 33.33%] ··· Running period.period_setitem.time_period_setitem                                                                                                             failed
[ 37.50%] ··· Running plotting.plot_timeseries_period.time_plot_timeseries_period                                                                                         308.25ms
[ 41.67%] ··· Running timeseries.timeseries_iter_periodindex.time_timeseries_iter_periodindex                                                                                6.57s
[ 45.83%] ··· Running timeseries.timeseries_iter_periodindex_preexit.time_timeseries_iter_periodindex_preexit                                                              55.71ms
[ 50.00%] ··· Running timeseries.timeseries_period_downsample_mean.time_timeseries_period_downsample_mean                                                                  26.72ms
[ 50.00%] · For pandas commit hash 768bf495:
[ 50.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....
[ 50.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 54.17%] ··· Running gil.nogil_datetime_fields.time_datetime_to_period                                                                                                    19.20ms
[ 58.33%] ··· Running gil.nogil_datetime_fields.time_period_to_datetime                                                                                                    51.12ms
[ 62.50%] ··· Running period.create_period_index_from_date_range.time_period_index                                                                                         54.20ms
[ 66.67%] ··· Running period.period_algorithm.time_period_index_drop_duplicates                                                                                            33.60μs
[ 70.83%] ··· Running period.period_algorithm.time_period_index_value_counts                                                                                                1.07ms
[ 75.00%] ··· Running period.period_algorithm.time_period_series_drop_duplicates                                                                                           16.89ms
[ 79.17%] ··· Running period.period_algorithm.time_period_series_value_counts                                                                                               6.42ms
[ 83.33%] ··· Running period.period_setitem.time_period_setitem                                                                                                             failed
[ 87.50%] ··· Running plotting.plot_timeseries_period.time_plot_timeseries_period                                                                                         294.17ms
[ 91.67%] ··· Running timeseries.timeseries_iter_periodindex.time_timeseries_iter_periodindex                                                                                4.01s
[ 95.83%] ··· Running timeseries.timeseries_iter_periodindex_preexit.time_timeseries_iter_periodindex_preexit                                                              40.75ms
[100.00%] ··· Running timeseries.timeseries_period_downsample_mean.time_timeseries_period_downsample_mean                                                                  24.24ms
    before     after       ratio
  [768bf495] [81262a57]
-   54.20ms    14.96ms      0.28  period.create_period_index_from_date_range.time_period_index

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.

```

By the way, I just realized that `period.period_setitem.time_period_setitem` fails even on `master`.

agraboso commented:
@jreback

jreback commented:
thanks!
